stellargraph/data/edge_data.py

```diff
@@ -0,0 +1,937 @@
+            Defaults to the constant DEFAULT_EDGE_WEIGHT.
+        num_implicit_edges (int, optional):
+            The number N of edges to be assigned implicit identifiers
+            in {0, 1, ..., N-1}; defaults to N = 0.
+        edge_ids (iterable, optional):
+            The collection of explicit edge identifiers to be indexed.
+        edge_types (iterable, optional):
+            The collection of explicit edge types to be summarised.
+
+    Note:
+        If any implicit edge identifiers are assigned, then these will
+        always be processed before any explicit edge identifiers.
+    """
+
+    def __init__(
+        self,
+        default_edge_type: Optional[Any] = None,
+        default_edge_weight: Optional[Number] = None,
+        num_implicit_edges: int = 0,
+        edge_ids: Optional[Iterable[Any]] = None,
+        edge_types: Optional[Iterable[Any]] = None,
+    ):
+        self._num_implicit_ids = _offset = num_implicit_edges
+        if edge_ids is None:
+            self._explicit_ids = {}
+        else:
+            self._explicit_ids = {
+                edge_id: (edge_idx + _offset)
+                for edge_idx, edge_id in enumerate(edge_ids)
+            }
+        _offset += len(self._explicit_ids)
+        super().__init__(default_edge_type, default_edge_weight, _offset)
+        if edge_types is None:
+            self._edge_types = set() if _offset == 0 else {self.default_edge_type()}
+        else:
+            self._edge_types = set(edge_types)
+
+    def is_identified(self) -> bool:
+        return self._num_implicit_ids == 0
+
+    def num_edges(self) -> int:
+        return self._num_implicit_ids + len(self._explicit_ids) + super().num_edges()
+
+    def edge_ids(self) -> Iterable[Any]:
+        return itertools.chain(
+            range(self._num_implicit_ids), self._explicit_ids.keys(), super().edge_ids()
+        )
+
+    def edge_index(self, edge_id: Any) -> int:
+        if isinstance(edge_id, int) and 0 <= edge_id < self._num_implicit_ids:
+            # Implicit identifier is the index
+            return edge_id
+        # Lookup explicit identifier
+        edge_idx = self._explicit_ids.get(edge_id, -1)
+        if edge_idx >= 0:
+            return edge_idx
+        # Lookup external identifier
+        return super().edge_index(edge_id)
+
+    def edge_type_set(self) -> Set:
+        return self._edge_types | super().edge_type_set()
+
+    def edges(self, include_info: bool = False) -> Iterable[tuple]:
+        if include_info:
+            return zip(
+                self.source_ids(),
+                self.target_ids(),
+                self.edge_ids(),
+                self.edge_types(),
+                self.edge_weights(),
+            )
+        return zip(self.source_ids(), self.target_ids())
+
+
+#############################################
+# Pandas data-frame of edge data:
+
+
+class PandasEdgeData(MappedEdgeData):
```

Comment:
I think that while we should have the Pandas edgelist to store edge features and edge types, we should have a typed adjacency list (MDAL*) that will be used for fast access to in and out neighbours of nodes.

As far as I can see this is done implictly by using the `EdgeCache`, but in my opinion this adds a layer of unnecessary complexity.  I like the fact that the edge cache will avoid the start-up time of building the complete adjacency list and allows to build only the subset of the adjacency list that is actually required. However, I feel strongly that having this general `EdgeData` class is too abstract.

Reply:
I'm a bit confused again (if ever I wasn't before), because #144 also talks about edges being a type-map. And if heterogeneous edge data is flattened to a Pandas data-frame (not clear if we would do this or the user), then we need a systematic way of determining which columns form the features for which type. This is what EdgeData is for.
Also, the caching of edges for efficiency, e.g. MDAL*, turns out to be separate from the specification of edge data. EdgeCache was designed to allow different implementations for different usages, e.g. heterogenous versus homogeneous. I'm still worried that a one-size-fits-all is going to be inefficient for simple data.

Reply:
> #144 also talks about edges being a type-map.

Yes, there is confusion here and I haven't been explaining my point of view well!  I agree with you that there needs to be three data structures: node features, edge features, and graph structure.

> and if heterogeneous edge data is flattened to a Pandas data-frame (not clear if we would do this or the user), then we need a systematic way of determining which columns form the features for which type.

I think I was wrong about this in https://github.com/stellargraph/stellargraph-ml-private-issues/issues/144,  I don't think we should flatten the array in our internal storage.  This was a potential way I though that users could want to specify the edge features, but let's focus on supporting the single unflattened format.

> Also, the caching of edges for efficiency, e.g. MDAL*, turns out to be separate from the specification of edge data

Agreed. The edge cache is the adjacency list for graph structure. 

> EdgeCache was designed to allow different implementations for different usages, e.g. heterogenous versus homogeneous. I'm still worried that a one-size-fits-all is going to be inefficient for simple data.

I did consider this, but I convinced myself that the overhead of looking up the type in the MDAL for a homogeneous graph would not be high.  However, I am not suggesting we have a one-size-fits all approach, but I'm thinking we can simplify things and for now have a single implementation and drop support for caching.